### PR TITLE
planner: do not push keep descending order to tiflash

### PR DESCRIPTION
**Pull request by lzmhhh123**
### What problem does this PR solve? 
As the title says.

### What is changed and how it works?
When do skyline pruning, check the properties for TiFlash access path.

### Check List 

Tests 

 - Unit test
 - Integration test

Side effects

 - Possible performance regression
 - Increased code complexity

### CODE REVIEW

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -428,6 +429,7 @@ func (ds *DataSource) setPreferredStoreType(hintInfo *tableHintInfo) {
 	}
 	if hintInfo.ifPreferTiFlash(alias) {
 		ds.preferStoreType |= preferTiFlash
+		ds.possibleAccessPaths = append(ds.possibleAccessPaths, &accessPath{isTablePath: true, storeType: kv.TiFlash})
```
Comment (lamxTyler) [inline planner/core/logical_plan_builder.go]: Can we do it when build possibleAccessPaths? From the function name, it should not change access paths.

Comment (lzmhhh123) [reply planner/core/logical_plan_builder.go]: If the `TiFlash` replica is not available. We can't generate a flash access path. But when the hint is set. we should enforce a flash access path.

Comment (lzmhhh123) [reply planner/core/logical_plan_builder.go]: @winoros That's not a temp behavior. For now, CBO can judge whether to choose the TiFlash access path by `tblInfo.TiFlashReplica.Available`. I hope when users set the hint, the planner enforce a TiFlash path without check the TiFlash replica status.

Comment (lamxTyler) [reply planner/core/logical_plan_builder.go]: Hint should not enforce something that tidb cannot do, it is only a suggestion for optimizer. Therefore, if the tiflash access path is not available, optimizer should not use it even if there are hints.

Comment (lzmhhh123) [reply planner/core/logical_plan_builder.go]: But we don't know whether the TiFlash access path is available. The table info of TiFlash replica is inaccurate information. For example, the user doesn't create the TiFlash replica by DDL or the respond of replica status is missed by the network. If the planner can be enforced by hint, then the TiFlash replica is always unavailable.

Comment (winoros) [reply planner/core/logical_plan_builder.go]: `the respond of replica status is missed by the network` This should not be a normal state?

Comment (lzmhhh123) [reply planner/core/logical_plan_builder.go]: Yes, it's a normal performance. But I still want to leave a function to force accessing the TiFlash. It's easy for both PD and TiFlash to check the bug when the replica is unavailable.

Comment (lamxTyler) [inline planner/core/logical_plan_builder.go]: Do we need to consider if the hints are conflicted? The path has already been added, but the preferStoreType is cleared.

File: planner/core/find_best_task.go [hunk 2]
```diff
@@ -460,10 +471,6 @@ func (ds *DataSource) findBestTask(prop *property.PhysicalProperty) (t task, err
 			}
 			continue
 		}
-		// TiFlash storage do not support index scan.
```
Comment (lamxTyler) [inline planner/core/find_best_task.go]: Why remove it?

File: planner/core/find_best_task.go [hunk 3]
```diff
@@ -430,6 +435,12 @@ func (ds *DataSource) findBestTask(prop *property.PhysicalProperty) (t task, err
 			}, nil
 		}
 		if path.isTablePath {
+			if ds.preferStoreType&preferTiFlash != 0 && path.storeType == kv.TiKV {
+				continue
+			}
+			if ds.preferStoreType&preferTiKV != 0 && path.storeType == kv.TiFlash {
+				continue
+			}
```
Comment (eurekaka) [inline planner/core/find_best_task.go]: We may end up with no plan generated if the hint specifies TiFlash while there is no TiFlash replica available?

File: planner/core/logical_plan_builder.go (hunk 1 above)
Comment (eurekaka) [reply planner/core/logical_plan_builder.go]: There would be 2 TiFlash paths then? since `getPossibleAccessPaths` may has appended one already.

Comment (lzmhhh123) [reply planner/core/logical_plan_builder.go]: Yes, but it doesn't affect the correctness. Just make the planner a bit slower.

File: planner/core/find_best_task.go (hunk 3 above)
Comment (lzmhhh123) [reply planner/core/find_best_task.go]: For now, if there is no TiFlash replica available, we still force a TiFlash plan.

File: planner/core/logical_plan_builder.go (hunk 1 above)
Comment (eurekaka) [reply planner/core/logical_plan_builder.go]: I prefer to check if TiFlash path exists before adding a new one. Also, please add comment to explain that this change is for testing purpose, and add a TODO to remind reverting it finally.

Comment (lzmhhh123) [reply planner/core/logical_plan_builder.go]: > I prefer to check if TiFlash path exists before adding a new one. Also, please add comment to explain that this change is for testing purpose, and add a TODO to remind reverting it finally.

@eurekaka Done.

**Pull request merged**